drivers/i3c/i3c_dw.c

```diff
@@ -0,0 +1,2378 @@
+	}
+
+	ret = clock_control_on(config->clock, NULL);
+	if (ret < 0) {
+		return ret;
+	}
+
+#ifdef CONFIG_I3C_USE_IBI
+	k_sem_init(&data->ibi_sts_sem, 0, 1);
+#endif
+	k_sem_init(&data->sem_xfer, 0, 1);
+	k_mutex_init(&data->mt);
+
+	data->mode = i3c_bus_mode(&config->common.dev_list);
+
+	/* reset all */
+	sys_write32(RESET_CTRL_ALL, config->regs + RESET_CTRL);
+
+	/* get DAT, DCT pointer */
+	data->datstartaddr =
+		DEVICE_ADDR_TABLE_ADDR(sys_read32(config->regs + DEVICE_ADDR_TABLE_POINTER));
+	data->dctstartaddr =
+		DEVICE_CHAR_TABLE_ADDR(sys_read32(config->regs + DEV_CHAR_TABLE_POINTER));
+
+	/* get max devices based on table depth */
+	data->maxdevs =
+		DEVICE_ADDR_TABLE_DEPTH(sys_read32(config->regs + DEVICE_ADDR_TABLE_POINTER));
+	data->free_pos = GENMASK(data->maxdevs - 1, 0);
+
+	/* get fifo sizes */
+	queue_capability = sys_read32(config->regs + QUEUE_SIZE_CAPABILITY);
+	data->txfifodepth = QUEUE_SIZE_CAPABILITY_TX_BUF_DWORD_SIZE(queue_capability);
+	data->rxfifodepth = QUEUE_SIZE_CAPABILITY_RX_BUF_DWORD_SIZE(queue_capability);
+	data->cmdfifodepth = QUEUE_SIZE_CAPABILITY_CMD_BUF_DWORD_SIZE(queue_capability);
+	data->respfifodepth = QUEUE_SIZE_CAPABILITY_RESP_BUF_DWORD_SIZE(queue_capability);
+	data->ibififodepth = QUEUE_SIZE_CAPABILITY_IBI_BUF_DWORD_SIZE(queue_capability);
+
+	/* get HDR capabilities */
+	ctrl_config->supported_hdr = 0;
+	hw_capabilities = sys_read32(config->regs + HW_CAPABILITY);
+	if (hw_capabilities & HW_CAPABILITY_HDR_TS_EN) {
+		ctrl_config->supported_hdr |= I3C_MSG_HDR_TSP | I3C_MSG_HDR_TSL;
+	}
+	if (hw_capabilities & HW_CAPABILITY_HDR_DDR_EN) {
+		ctrl_config->supported_hdr |= I3C_MSG_HDR_DDR;
+	}
+
+	/* if the boot condition starts as a target, then it's a secondary controller */
+	device_ctrl_ext = sys_read32(config->regs + DEVICE_CTRL_EXTENDED);
+	if (DEVICE_CTRL_EXTENDED_DEV_OPERATION_MODE(device_ctrl_ext) &
+	    DEVICE_CTRL_EXTENDED_DEV_OPERATION_MODE_SLAVE) {
+		ctrl_config->is_secondary = true;
+	} else {
+		ctrl_config->is_secondary = false;
+	}
+
+	ret = init_scl_timing(dev);
+	if (ret != 0) {
+		return ret;
+	}
+
+	enable_intr(dev);
+
+	/* disable ibi */
+	sys_write32(IBI_REQ_REJECT_ALL, config->regs + IBI_SIR_REQ_REJECT);
+	sys_write32(IBI_REQ_REJECT_ALL, config->regs + IBI_MR_REQ_REJECT);
+
+	/* disable hot-join */
+	sys_write32(sys_read32(config->regs + DEVICE_CTRL) | (DEV_CTRL_HOT_JOIN_NACK),
+		    config->regs + DEVICE_CTRL);
+
+	ret = i3c_addr_slots_init(dev);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = set_controller_info(dev);
+	if (ret) {
+		return ret;
+	}
+
+	dw_i3c_enable_controller(config, true);
+
+	if (!(ctrl_config->is_secondary)) {
+		/* Perform bus initialization */
+		ret = i3c_bus_init(dev, &config->common.dev_list);
+		/* Bus Initialization Complete, allow HJ ACKs */
+		sys_write32(sys_read32(config->regs + DEVICE_CTRL) & ~(DEV_CTRL_HOT_JOIN_NACK),
+			    config->regs + DEVICE_CTRL);
+	}
+
+	return 0;
+}
+
+static const struct i3c_driver_api dw_i3c_api = {
```

Comment:
Wrap the API using the `DEVICE_API` macro.

```suggestion
static DEVICE_API(i3c, dw_i3c_api) = {
```

Reply:
done